text/0000-scoped-take-2.md

```diff
@@ -0,0 +1,262 @@
+does not guarantee that the destructor for a value that escapes will be run.
+
+Note that this does *not* mean that RAII should be avoided in general. **RAII is
+still a safe, idiomatic, and ergonomic style of programming for Rust.** It just
+means that you cannot write `unsafe` code whose safety relies on RAII to run
+destructors.
+
+It may be possible to finesse this issue by locating every possible source of
+destructor leakage (such as `Rc` cycles) and tying it to some kind of "`Leak`"
+marker trait, which would be implemented by default for all Rust types (and
+basically means "safe to leak").  Ideas along these lines have been discussed
+extensively [elsewhere](https://github.com/rust-lang/rfcs/pull/1066), but they
+represent longer-term language changes that may or may not work out.
+
+This RFC, by contrast, proposes a general API that works in today's Rust, and
+can:
+
+* Guarantee execution of a piece of code prior to a scope being exited, even in
+  the presence of unwinding, `Rc` cycles, and so on.
+
+* Thereby be used to recover a safe `thread::scoped` API.
+
+While the API proposed here is slightly less ergonomic than the RAII-style
+`scoped` API, **it has the benefit of much more clearly marking the point at
+which children threads will be joined, which was left more implicit in the old
+API.**
+
+# Detailed design
+
+The proposal has two pieces: a general mechanism for "deferred computation",
+and a new `thread::scoped` API that takes advantage of it.
+
+## Deferred computation
+
+Here's a very simple API for deferring computation:
+
+```
+// in std:
+
+mod thread {
+    pub struct Scope<'a> { ... }
+
+    pub fn scope<'a, F, R>(f: F) -> R where F: FnOnce(&Scope<'a>) -> R;
+
+    impl<'a> Scope<'a> {
+        pub fn defer<F>(&self, f: F) where F: FnOnce() + 'a;
+    }
+}
+```
+
+(This is put in the `std::thread` module because (1) scopes are per-thread
+concepts (2) it fits nicely with other functionality like `panicking` and
+`catch_panic`.)
+
+You call `scope` to introduce a new `Scope` value, which is passed into a
+closure that is immediately invoked.
+
+The closure can use the `defer` method to register callbacks to invoke upon
+*any* exit from the callback, including unwinding. Unlike RAII guards, there is
+no way to leak these callbacks, and the implementation shown below works around
+[known cases](https://github.com/rust-lang/rust/issues/14875) of destructor
+leakage.
+
+This is a generally useful mechanism that can avoid the need to create custom
+RAII guards for situations where you might use `try`/`finally` in other
+languages. But it is also just the support needed for scoped threads.
+
+## Scoped threads
+
+To recover scoped threads, we extend the `Scope` type with a method for spawning
+threads:
+
+```rust
+impl<'defer> Scope<'defer> {
+    pub fn spawn<'body, F, T>(&self, f: F) -> thread::JoinHandle<T> where
+        'defer: 'body,   // anything borrowed by the body must outlive the Scope
```

Comment:
@kballard `'defer: 'body` means `'defer` *outlives* `'body`. That means it's safe to use an `&'defer` in place of an `&'body`. So the relationship is covariant the way it is expressed. This is somewhat arbitrary though, since lifetimes can't be used in isolation. The only subtyping relation that can really exist for them is in the variance of types using them, so it made sense to make their "subtyping" covariant with how they're typically used.